### PR TITLE
Save 1MB by not using charmbracelet/log

### PR DESCRIPTION
**Pull request by Mic92**
Adding nixos-facter tipped us over the edge for 1GB RAM in nixos-anywhere. Another upside is that less dependencies means less updates.

### CODE REVIEW

File: pkg/virt/detect_vm_xen.go [hunk 1]
```diff
@@ -47,7 +47,7 @@ func detectXenDom0() (bool, error) {
 		}
 
 		result = (features & (1 << xenFeatDom0)) != 0
-		l.Debugf("found %s with value %x", featuresPath, features)
+		l.Debug("found %s with value %x", featuresPath, features)
```
Comment (Mic92) [inline pkg/virt/detect_vm_xen.go:50]: Need to check this one again.

File: pkg/virt/detect_vm_xen.go [hunk 2]
```diff
@@ -59,13 +59,13 @@ func detectXenDom0() (bool, error) {
...
 		// must be running as a Xen guest
-		l.Debugf("%s does not exist, DomU is indicated", capabilitiesPath)
+		l.Debug("capabilities path does not exist, DomU is indicated", "path", capabilitiesPath)
 		return false, nil
 	} else if err != nil {
 		return false, err
 	}
 
-	l.Debugf("found %s with value %s", capabilitiesPath, string(b))
+	l.Debug("found %s with value %s", capabilitiesPath, string(b))
```
Comment (Mic92) [inline pkg/virt/detect_vm_xen.go:68]: And here

**Comment by Mic92** [comment]
Now we are at half the binary size that we started with.

**Comment by brianmcgee** [comment]
I hope you started by trying to rip things out of `nixos-anywhere` before labotomizing `nixos-facter` 😂 

To be fair we didn't really need the log and flag features, they're my go to when starting out.

What kinda max binary size do we have to work with if we aren't gonna blow nixos-anywhere's budget in the future?

**Comment by Mic92** [comment]
> I hope you started by trying to rip things out of `nixos-anywhere` before labotomizing `nixos-facter` 😂

Oh yeah, we did optimize hard: https://github.com/nix-community/nixos-images/blob/8650b1df8f965bf33dec79351646e19406ff9d50/nix/noninteractive.nix#L17

https://github.com/nix-community/nixos-images/blob/8650b1df8f965bf33dec79351646e19406ff9d50/nix/zfs-minimal.nix#L6

Ideally we could get rid of perl... but there is still nixos-generate-config.

> 
> To be fair we didn't really need the log and flag features, they're my go to when starting out.
> 
> What kinda max binary size do we have to work with if we aren't gonna blow nixos-anywhere's budget in the future?

It's only after I added nixos-facter, when nixos tests starting to go out-of-memory. It might have just put it over the edge.

**Comment by brianmcgee** [comment]
These changes take the binary from `5M` down to `2.6M`. 

I'm mindful of #117, which I'll look at in the next day or two and consider what changes it might require. 

Is there an easy way to test if I've blown the `nixos-anywhere` RAM budget? A representative test to run or something?

**Comment by brianmcgee** [comment]
> It's only after I added nixos-facter, when nixos tests starting to go out-of-memory. It might have just put it over the edge. I have to see when we get back into the safe zone.

Just re-read this and realised you're still unsure about what the budget might be. I'll let you keep working through it.

**Comment by Mic92** [comment]
I just gave us more breathing space by overriding python with python minimal in various filesystem tools. I would say libraries that add hardware detection should be fine. In the case of lsusb, I briefly studied the code, and I think it would be hard to find a library that handles all the edge cases that it handles. So I would instead suggesting parsing the output of lsusb -v. But in any case having a small binary will make it an easier sell later when we want to replace nixos-generate-config in NixOS as well.

**Comment by Mic92** [comment]
This now is an easier to read `--help`:

```
nixos-facter [flags]
Hardware report generator

Usage:
  nixos-facter [flags]

Flags:
  --ephemeral          capture all ephemeral properties e.g. swap, filesystems and so on
  -h, --help           help for nixos-facter
  -o, --output string  path to write the report
  --swap               capture swap entries
  --hardware strings   Hardware items to probe.
                       Default: memory,pci,net,serial,cpu,bios,monitor,scsi,usb,prom,sbus,sys,sysfs,udev,block,wlan
                       Possible values: memory,pci,isapnp,net,floppy,misc,misc_serial,misc_par,misc_floppy,serial,cpu,bios,monitor,mouse,scsi,usb,usb_mods,adb,modem,modem_usb,parallel,parallel_lp,parallel_zip,isa,isa_isdn,isdn,kbd,prom,sbus,braille,braille_alva,braille_fhp,braille_ht,ignx11,sys,bios_vbe,isapnp_old,isapnp_new,isapnp_mod,braille_baum,manual,fb,veth,pppoe,scan,pcmcia,fork,parallel_imm,s390,cpuemu,sysfs,s390disks,udev,block,block_cdrom,block_part,edd,edd_mod,bios_ddc,bios_fb,bios_mode,input,block_mods,bios_vesa,cpuemu_debug,scsi_noserial,wlan,bios_crc,hal,bios_vram,bios_acpi,bios_ddc_ports,modules_pata,net_eeprom,x86emu,max,lxrc,all
```

**Review by brianmcgee — APPROVED** [review]

**Comment by brianmcgee** [comment]
> I briefly studied the code, and I think it would be hard to find a library that handles all the edge cases that it handles. So I would instead suggesting parsing the output of lsusb -v. But in any case having a small binary will make it an easier sell later when we want to replace nixos-generate-config in NixOS as well.

Good to know. I'm going to give this some focused attention today or tomorrow.

**Pull request merged**